File: Java/src/Graph/Topological.java
package Graph;

import edu.princeton.cs.algs4.StdOut;
import edu.princeton.cs.algs4.SymbolDigraph;
import edu.princeton.cs.algs4.Digraph;
import edu.princeton.cs.algs4.DirectedCycle;
import edu.princeton.cs.algs4.DepthFirstOrder;

/**
 * 拓扑排序
 *
 * @author igaozp
 * @since 2017-07-14
 * @version 1.0
 */
public class Topological {
    /**
     * 拓扑序列
     */
    private Iterable<Integer> order;

    /**
     * 拓扑排序的构造方法
     *
     * @param G 有向图
     */
    public Topological(Digraph G) {
        DirectedCycle cycleFinder = new DirectedCycle(G);

        // 检查是否有环，若没有则进行拓扑排序
        if (!cycleFinder.hasCycle()) {
            DepthFirstOrder dfs = new DepthFirstOrder(G);
            order = dfs.reversePost();
        }
    }

    /**
     * 获取拓扑序列
     *
     * @return 拓扑序列
     */
    public Iterable<Integer> order() {
        return order;
    }

    /**
     * 检查是否是有向无环图
     *
     * @return {@code true} 是有向无环图
     *         {@code false} 不是有向无环图
     */
    public boolean isDAG() {
        return order != null;
    }

    public static void main(String[] args) {
        String filename = args[0];
        String separator = args[0];
        SymbolDigraph sg = new SymbolDigraph(filename, separator);
        Topological top = new Topological(sg.G());

        for (int v : top.order()) {
            StdOut.println(sg.name(v));
        }
    }
}

File: Kotlin/src/me/metatom/kotlin/graph/Topological.kt
package me.metatom.kotlin.graph

import edu.princeton.cs.algs4.*
import edu.princeton.cs.algs4.DepthFirstOrder
import edu.princeton.cs.algs4.Digraph
import edu.princeton.cs.algs4.DirectedCycle
import edu.princeton.cs.algs4.EdgeWeightedDigraph
import edu.princeton.cs.algs4.Topological

/**
 * 拓扑排序
 *
 * @author igaozp
 * @since 2017-09-11
 * @version 1.0
 */
class Topological {
    // 拓扑序列
    private var order: Iterable<Int>? = null
    // rank[v] = rank of vertex v in order
    private var rank = MutableList(0) { 0 }

    /**
     * 拓扑排序的构造方法
     *
     * @param G 有向图
     */
    constructor(G: Digraph) {
        val cycleFinder = DirectedCycle(G)

        // 检查是否有环，若没有则进行拓扑排序
        if (!cycleFinder.hasCycle()) {
            val dfs = DepthFirstOrder(G)
            order = dfs.reversePost()
            var i = 0
            for (v in order!!) {
                rank[v] = i++
            }
        }
    }

    /**
     * 构造方法
     *
     * @param G 带权重的有向图
     */
    constructor(G: EdgeWeightedDigraph) {
        val finder = EdgeWeightedDirectedCycle(G)
        if (!finder.hasCycle()) {
            val dfs = DepthFirstOrder(G)
            order = dfs.reversePost()
        }
    }

    /**
     * 获取拓扑序列
     *
     * @return 拓扑序列
     */
    fun order(): Iterable<Int>? = if (order != null) order else null

    /**
     * 检查是否是有向无环图
     *
     * @return `true` 是有向无环图
     *         `false` 不是有向无环图
     */
    fun isDAG(): Boolean = order != null

    /**
     * 验证顶点
     *
     * @param v 需要验证的顶点
     */
    private fun validateVertex(v: Int) {
        if (v < 0 || v >= rank.size) {
            throw IllegalArgumentException("vertex $v is not between 0 and ${rank.size - 1}")
        }
    }
}

/**
 *  单元测试
 */
fun main(args: Array<String>) {
    val filename = args[0]
    val separator = args[0]
    val sg = SymbolDigraph(filename, separator)
    val top = Topological(sg.G())

    for (v in top.order()) {
        StdOut.println(sg.name(v))
    }
}